From b8b3286ec3c6e5ebb9117173c17ccbe4e2cd8958 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Mon, 9 Dec 2024 17:49:32 +0100
Subject: [PATCH 86/86] kernel32/tests: Add SleepEx to finish overlapped read.

=================================================================
==508==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7ffffe1ea7f8 at pc 0x0001401a4933 bp 0x7ffffe1e2a80 sp 0x7ffffe1e2ac8
READ of size 8 at 0x7ffffe1ea7f8 thread T0
0208:fixme:file:server_get_file_info Unsupported info class e
    #0 0x0001401a4932 in FileIOComplete /home/bernhard/data/entwicklung/2024/wine\wine/dlls/kernel32/tests/file.c:3337:24
    #1 0x6ffffacef75f in KiUserApcDispatcher (C:\windows\system32\ntdll.dll+0x17000f75f)
    #2 0x6ffffacedf53 in NtDelayExecution (C:\windows\system32\ntdll.dll+0x17000df53)
    #3 0x6ffffa6eec5f in SleepEx /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernelbase\sync.c:346:14
    #4 0x00014015a0e3 in test_read_write /home/bernhard/data/entwicklung/2024/wine\wine/dlls/kernel32/tests/file.c:3451:5
    #5 0x00014015a0e3 in func_file /home/bernhard/data/entwicklung/2024/wine\wine/dlls/kernel32/tests/file.c:6348:5
    #6 0x0001404d7fa1 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #7 0x0001404d7fa1 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #8 0x0001404d9f8f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #9 0x6ffffbfd4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #10 0x6ffffacefa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ffffe1ea7f8 is located in stack of thread T0 at offset 30232 in frame
    #0 0x00014013223f in func_file /home/bernhard/data/entwicklung/2024/wine\wine/dlls/kernel32/tests/file.c:6296

  This frame has 205 object(s):
    [32, 292) 'cwd.i6426' (line 6001)
    [368, 628) 'temp_dir.i6427' (line 6001)
    [704, 964) 'cwd.i' (line 5922)
    [1040, 1300) 'temp_dir.i' (line 5922)
    [1376, 1576) 'name_buffer.i' (line 5922)
    [1648, 1668) 'buffer.i6266' (line 5922)
    [1712, 1728) 'io.i6267' (line 5924)
    [1744, 1748) 'size.i6268' (line 5927)
    [1760, 2020) 'path.i6021' (line 4751)
    [2096, 2356) 'filename.i6022' (line 4751)
    [2432, 2436) 'buffer.i6023' (line 4751)
    [2448, 2452) 'size.i6024' (line 4754)
    [2464, 2984) 'path.i5885' (line 5880)
    [3120, 3640) 'temp_path.i5886' (line 5880)
    [3776, 3784) 'ft1.i5887' (line 5881)
    [3808, 3816) 'ft2.i5888' (line 5881)
    [3840, 3844) 'len.i5889' (line 5882)
    [3856, 3878) 'path.i5864' (line 5859)
    [3920, 4520) 'data.i5865' (line 5862)
    [4656, 4916) 'file_name.i5674' (line 5792)
    [4992, 5252) 'file_name2.i' (line 5792)
    [5328, 5588) 'temp_path.i5675' (line 5793)
    [5664, 5924) 'temp_path.i5568' (line 5730)
    [6000, 6260) 'file_name.i5569' (line 5731)
    [6336, 6340) 'bytes_count.i' (line 5732)
    [6352, 6384) 'ov.i5570' (line 5733)
    [6416, 6448) 'ovl.i5347' (line 5618)
    [6480, 6512) 'ovl2.i' (line 5618)
    [6544, 6552) 'povl.i5348' (line 5618)
    [6576, 6640) 'entries.i' (line 5619)
    [6672, 6680) 'key.i5349' (line 5620)
    [6704, 6708) 'count.i5350' (line 5622)
    [6720, 6724) 'size.i5351' (line 5623)
    [6736, 6772) 'info.i' (line 5597)
    [6816, 7336) 'tempFileFrom.i' (line 5518)
    [7472, 7992) 'tempFileTo1.i' (line 5518)
    [8128, 8648) 'tempFileTo2.i' (line 5518)
    [8784, 9304) 'tempPath.i5155' (line 5519)
    [9440, 9448) 'fileattrinfo.i' (line 5401)
    [9472, 9588) 'protinfo.i' (line 5402)
    [9632, 9656) 'stdinfo.i' (line 5403)
    [9696, 9712) 'compressinfo.i' (line 5404)
    [9728, 9729) 'dispinfo.i4919' (line 5405)
    [9744, 9748) 'hintinfo.i' (line 5406)
    [9760, 9800) 'basicinfo.i' (line 5407)
    [9840, 10100) 'tempFileName.i4920' (line 5408)
    [10176, 10436) 'tempPath.i4921' (line 5409)
    [10512, 11032) 'temp_path.i4625' (line 5281)
    [11168, 11688) 'test_path.i4626' (line 5281)
    [11824, 12344) 'long_path.i4627' (line 5282)
    [12480, 13000) 'result_path.i4628' (line 5282)
    [13136, 13676) 'dos_path.i4629' (line 5283)
    [13808, 14328) 'drive_part.i' (line 5284)
    [14464, 15084) 'volume_path.i' (line 5286)
    [15216, 16256) 'nt_path.i' (line 5287)
    [16384, 16644) 'temp_path.i4444' (line 5193)
    [16720, 16980) 'test_path.i' (line 5193)
    [17056, 17316) 'long_path.i' (line 5194)
    [17392, 17652) 'result_path.i' (line 5194)
    [17728, 17993) 'dos_path.i' (line 5195)
    [18064, 18324) 'path.i4319' (line 5101)
    [18400, 18660) 'fname.i' (line 5101)
    [18736, 18752) 'buf.i4320' (line 5102)
    [18768, 18776) 'hdup.i' (line 5103)
    [18800, 18804) 'bytes.i4321' (line 5104)
    [18816, 19076) 'temp_path.i3961' (line 4846)
    [19152, 19412) 'filename.i3962' (line 4846)
    [19488, 19492) 'size.i' (line 4848)
    [19504, 19508) 'tx.i' (line 4848)
    [19520, 19528) 'key.i' (line 4849)
    [19552, 19568) 'fse.i' (line 4850)
    [19584, 19616) 'ovl.i3963' (line 4851)
    [19648, 19656) 'povl.i' (line 4851)
    [19680, 19728) 'si.i' (line 4852)
    [19760, 19764) 'count.i3600' (line 4616)
    [19776, 20036) 'path.i' (line 4617)
    [20112, 20372) 'filename.i3601' (line 4617)
    [20448, 20464) 'privs.i' (line 4618)
    [20480, 20488) 'token.i' (line 4619)
    [20512, 20772) 'tempPath.i3481' (line 4510)
    [20848, 21108) 'tempFileName.i3482' (line 4510)
    [21184, 21440) 'buffer.i3483' (line 4510)
    [21504, 21760) 'tickCount.i' (line 4510)
    [21824, 22344) 'tempFileNameW.i' (line 4511)
    [22480, 22484) 'count.i3484' (line 4513)
    [22496, 22520) 'fileIdDescr.i' (line 4516)
    [22560, 22820) 'tempPath.i' (line 4345)
    [22896, 23156) 'tempFileName.i' (line 4345)
    [23232, 24256) 'buffer.i3172' (line 4345)
    [24384, 24388) 'written.i3173' (line 4347)
    [24400, 24404) 'priohintinfo.i' (line 4354)
    [24416, 24424) 'allocinfo.i' (line 4355)
    [24448, 24449) 'dispinfo.i' (line 4356)
    [24464, 24472) 'eofinfo.i' (line 4357)
    [24496, 24520) 'renameinfo.i' (line 4358)
    [24560, 25080) 'replaced.i' (line 4126)
    [25216, 25736) 'replacement.i' (line 4126)
    [25872, 26392) 'backup.i' (line 4126)
    [26528, 27048) 'temp_path.i3030' (line 4128)
    [27184, 27193) 'directory.i' (line 3811)
    [27216, 27248) 'ov.i2847' (line 3733)
    [27280, 27284) 'result.i2848' (line 3734)
    [27296, 27432) 'ofs.i' (line 3551)
    [27504, 27764) 'buff.i' (line 3563)
    [27840, 28880) 'buff_long.i' (line 3564)
    [29008, 29136) 'filled_0xA5.i' (line 3565)
    [29168, 29172) 'bytes.i' (line 3392)
    [29184, 29188) 'old_prot.i' (line 3392)
    [29200, 29460) 'temp_path.i2303' (line 3394)
    [29536, 29796) 'filename.i2304' (line 3395)
    [29872, 30132) 'szFile.i' (line 3343)
    [30208, 30240) 'ovl.i' (line 3347) <== Memory access at offset 30232 is inside this variable
    [30272, 30276) 'count.i' (line 3364)
    [30288, 30296) 'h.i' (line 3246)
    [30320, 30328) 'h2.i' (line 3246)
    [30352, 30384) 'ov.i' (line 2212)
    [30416, 30420) 'done.i' (line 2213)
    [30432, 30688) 'buf.i1791' (line 2215)
    [30752, 30757) 'pattern.i' (line 2215)
    [30784, 31044) 'temp_path.i1792' (line 2217)
    [31120, 31380) 'temp_fname.i' (line 2217)
    [31456, 31460) 'written.i1596' (line 2285)
    [31472, 31504) 'overlapped.i' (line 2286)
    [31536, 31856) 'find_data.i1553' (line 3034)
    [31920, 32432) 'correct.i' (line 3145)
    [32496, 33008) 'incorrect.i' (line 3146)
    [33072, 33584) 'missing.i' (line 3147)
    [33648, 33664) 'quoted.i' (line 3157)
    [33680, 33940) 'windowsdir.i.i1529' (line 2705)
    [34016, 34336) 'search_results.i1530' (line 2913)
    [34400, 34405) 'buffer.i1531' (line 2915)
    [34432, 34692) 'windowsdir.i.i' (line 2705)
    [34768, 35088) 'data.i' (line 2713)
    [35152, 35157) 'buffer.i1404' (line 2715)
    [35184, 35284) 'buffer2.i' (line 2716)
    [35328, 35588) 'nonexistent.i' (line 2717)
    [35664, 35924) 'tmp.i' (line 2779)
    [36000, 36520) 'temp_path.i1307' (line 2177)
    [36656, 37176) 'source.i1308' (line 2178)
    [37312, 37832) 'dest.i1309' (line 2178)
    [37968, 38228) 'tempdir.i' (line 1996)
    [38304, 38564) 'source.i1162' (line 1997)
    [38640, 38900) 'dest.i1163' (line 1997)
    [38976, 39296) 'find_data.i' (line 1999)
    [39360, 39364) 'ret.i' (line 2002)
    [39376, 39696) 'fd.i' (line 2146)
    [39760, 40020) 'temppath.i' (line 2147)
    [40096, 40616) 'pathW.i' (line 1943)
    [40752, 41272) 'pathsubW.i' (line 1944)
    [41408, 41668) 'temp_path.i945' (line 1867)
    [41744, 42004) 'temp_file.i' (line 1867)
    [42080, 42088) 'argv.i' (line 1869)
    [42112, 42632) 'temp_path.i875' (line 1717)
    [42768, 43288) 'filename.i876' (line 1718)
    [43424, 43456) 'exparams.i' (line 1719)
    [43488, 44008) 'temp_path.i812' (line 1628)
    [44144, 44664) 'filename.i813' (line 1629)
    [44800, 45060) 'temp_path.i609' (line 1271)
    [45136, 45396) 'dirname.i' (line 1271)
    [45472, 45732) 'filename.i' (line 1272)
    [45808, 46068) 'windowsdir.i610' (line 1274)
    [46144, 46404) 'Volume_1.i' (line 1275)
    [46480, 46992) 'buffer.i611' (line 1276)
    [47056, 47060) 'len.i612' (line 1279)
    [47072, 47124) 'Finfo.i' (line 1297)
    [47168, 47688) 'curdir.i' (line 1298)
    [47824, 48084) 'temp_path.i556' (line 4244)
    [48160, 48420) 'file_name.i' (line 4245)
    [48496, 48500) 'written.i' (line 4246)
    [48512, 48772) 'temp_path.i480' (line 1193)
    [48848, 49108) 'source.i481' (line 1194)
    [49184, 49444) 'dest.i482' (line 1194)
    [49520, 50040) 'source.i416' (line 957)
    [50176, 50696) 'dest.i417' (line 957)
    [50832, 51352) 'temp_path.i418' (line 957)
    [51488, 51520) 'params.i' (line 958)
    [51552, 51560) 'ft1.i' (line 960)
    [51584, 51592) 'ft2.i' (line 960)
    [51616, 51620) 'len.i' (line 961)
    [51632, 51642) 'buf.i' (line 962)
    [51664, 52184) 'temp_path.i' (line 911)
    [52320, 52840) 'source.i' (line 912)
    [52976, 53496) 'dest.i' (line 912)
    [53632, 53892) 'out.i' (line 1830)
    [53968, 54238) 'expected.i' (line 1831)
    [54304, 54574) 'windowsdir.i' (line 1832)
    [54640, 54643) 'windowsdrive.i' (line 1833)
    [54656, 64656) 'buffer.i289' (line 596)
    [64912, 64913) 'checksum.i290' (line 603)
    [64928, 74928) 'buffer.i245' (line 551)
    [75184, 85184) 'buffer.i212' (line 512)
    [85440, 85441) 'buffer.i141' (line 474)
    [85456, 85472) 'strW.i.i' (line 277)
    [85488, 85504) 'str.i.i' (line 278)
    [85520, 85536) 'filenameW.i' (line 295)
    [85552, 85600) 'attr.i' (line 296)
    [85632, 85648) 'io.i' (line 297)
    [85664, 95664) 'buffer.i103' (line 299)
    [95920, 96240) 'search_results.i' (line 300)
    [96304, 96312) 'slashname.i' (line 301)
    [96336, 96344) 'file.i' (line 303)
    [96368, 106368) 'buffer.i55' (line 170)
    [106624, 106625) 'checksum.i' (line 177)
    [106640, 116640) 'buffer.i' (line 123)
    [116896, 117156) 'temp_path' (line 6297)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-scope /home/bernhard/data/entwicklung/2024/wine\wine/dlls/kernel32/tests/file.c:3337:24 in FileIOComplete
Shadow bytes around the buggy address:
  0x7ffffe1ea500: f2 f2 f2 f2 f2 f2 f2 f2 00 00 00 00 00 00 00 00
  0x7ffffe1ea580: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ea600: 00 00 00 00 00 00 00 00 04 f2 f2 f2 f2 f2 f2 f2
  0x7ffffe1ea680: f2 f2 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1ea700: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
=>0x7ffffe1ea780: f8 f8 f8 f2 f2 f2 f2 f2 f2 f2 f2 f2 f8 f8 f8[f8]
  0x7ffffe1ea800: f2 f2 f2 f2 f8 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f8
  0x7ffffe1ea880: f8 f8 f2 f2 f2 f2 f8 f2 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1ea900: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1ea980: f8 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 f2 f2 f2 f2
  0x7ffffe1eaa00: f8 f2 f2 f2 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==508==ABORTING
0208:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:201808: dlls/kernel32/tests/x86_64-windows/file.ok] Fehler 1
---
 dlls/kernel32/tests/file.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/kernel32/tests/file.c b/dlls/kernel32/tests/file.c
index c57edebb316..33357e27356 100644
--- a/dlls/kernel32/tests/file.c
+++ b/dlls/kernel32/tests/file.c
@@ -3378,6 +3378,7 @@ static void test_async_file_errors(void)
     }
     ok(completion_count == 0, "completion routine should only be called when ReadFileEx succeeds (this rule was violated %d times)\n", completion_count);
     /*printf("Error = %ld\n", GetLastError());*/
+    SleepEx(0, TRUE); /* Without it ASan would trigger a stack-use-after-scope in SleepEx in test_read_write later. */
     HeapFree(GetProcessHeap(), 0, lpBuffer);
 }
 
-- 
2.47.1

